quickwit/quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -595,6 +626,37 @@ fn spawn_consumer_poll_loop(
     })
 }
 
+/// Attempts to drain the overflow buffer by sending as many events as possible to the channel
+/// `events_tx` before `max_duration` has passed.
+fn drain_overflow_buffer<T>(
+    overflow_buffer: &mut VecDeque<T>,
+    events_tx: &mpsc::Sender<T>,
+    max_duration: Duration,
+) {
+    if overflow_buffer.is_empty() {
+        return;
+    }
+    let now = tokio::time::Instant::now();
+    let deadline = now + max_duration;
+    Handle::current().block_on(async {
+        while let Some(event) = overflow_buffer.pop_back() {
+            let permit = match tokio::time::timeout_at(deadline, events_tx.reserve()).await {
```

Comment:
`.send_timeout()` would have been simpler I think?

The error returns the value so you can push it back in the overflow buffer.